source/retryable-reads/retryable-reads.rst

```diff
@@ -0,0 +1,644 @@
+reads are retried exactly once.
+
+Motivation for Change 
+======================
+
+Drivers currently have an API for the retryability of write operations but not
+for read operations. The driver API needs to be extended to include support for
+retryable behavior for read operations.
+
+Design Rationale 
+=================
+
+The design of this specification is based off the `Retryable Writes
+specification
+<https://github.com/mongodb/specifications/blob/master/source/retryable-writes/retryable-writes.rst#design-rationale>`__.
+It modifies the driver API as little as possible to introduce the concept
+retryable behavior for read operations.
+
+Alternative retry strategies (e.g. exponential back-off, incremental intervals,
+regular intervals, immediate retry, randomization) were considered, but the
+behavior of a single, immediate retry attempt was chosen in the interests of
+simplicity as well as consistency with the design for retryable writes.
+
+See the `future work`_ section for potential upcoming changes
+to retry mechanics.
+
+Backwards Compatibility 
+========================
+
+The API changes to support retryable reads extend the existing API but do not
+introduce any backward breaking changes. Existing programs that do not make use
+of retryable reads will continue to compile and run correctly.
+
+N.B.: Applications with custom retry logic that choose to enable retryable reads
+may need to redo their custom retry logic to ensure that the reads are retried
+as desired. e.g. if an application has custom logic that retries reads n times
+and enables retryable reads, then the application could end up retrying reads up
+to 2n times.
+
+The note above will also apply if an application upgrades to a version of the
+driver where that defaults to enabling retryable reads.
+
+Reference Implementation 
+=========================
+
+The C# and Python drivers will provide the reference implementations. See
+`CSHARP-2429 <https://jira.mongodb.org/browse/CSHARP-2429>`__ and `PYTHON-1674
+<https://jira.mongodb.org/browse/PYTHON-1674>`__.
+
+Security Implications
+=====================
+
+None.
+
+Future work
+===========
+
+1. A later specification may allow operations (including read) to be retried any
+   number of times during a singular timeout period.
+
+2. A later specification may have `drivers default to enabling retryable writes
+   <https://jira.mongodb.org/browse/WRITING-3115>`__, in which case, drivers
+   MUST also default to enabling retryable reads.
+
+3. Any future changes to the the applicable parts of `retryable writes
+   specification
+   <https://github.com/mongodb/specifications/blob/master/source/retryable-writes/>`__
+   may also need to be reflected in the retryable reads specification, and vice
+   versa.
+
+4. We may revisit the decision not retry ``Cursor.getMore()`` (see `Q&A`_).
+
+5. To improve performance on servers without “Early Failure on Socket
```

Comment:
it's not crazy, but probably not what you want.

1. you'll need to burn the session you've killed.  I'd have to think further, but I worry that after returning from kill, you may still have bits of it running (if there are any errors).
2. it will take a while.  a kill has to talk to every non-config mongod in the cluster (i.e. all primaries, all secondaries, etc, of every shard).  In addition, in order to protect the system against getting overloaded with these requests, every server allows no more than one killsession operation at a time.  Operations that attempt to killsessions while a killsession is running are batched together and run simultaneously after the current one finishes

If the point of retryable reads is partially to allow you to retarget faster, that may not get you what you want

Reply:
Thanks Jason for chiming in :) Given the uncertainty, I can remove this text and the potential performance problem can be re-addressed in future only  if it's needed. @ShaneHarvey , are you comfortable with this?

Reply:
I would like to see this idea be incorporated into the "Design Rationale" section. Maybe under a "rejected designs" section. This is useful information that we will want to refer back to when considering the client side operation timeout feature.